Heap/Heap.swift

```diff
@@ -4,58 +4,55 @@
 //
 
 public struct Heap<T> {
+  
   /** The array that stores the heap's nodes. */
-  var elements = [T]()
+  private(set) var nodes = [T]()
   
-  /** Determines whether this is a max-heap (>) or min-heap (<). */
-  fileprivate var isOrderedBefore: (T, T) -> Bool
+  /**
+   * Determines how to compare two nodes in the heap.
+   * Use '>' for a max-heap or '<' for a min-heap,
+   * or provide a comparing method if the heap is made
+   * of custom elements, for example tuples.
+   */
+  private var orderCriteria: (T, T) -> Bool
   
   /**
    * Creates an empty heap.
    * The sort function determines whether this is a min-heap or max-heap.
-   * For integers, > makes a max-heap, < makes a min-heap.
+   * For comparable data types, > makes a max-heap, < makes a min-heap.
    */
   public init(sort: @escaping (T, T) -> Bool) {
-    self.isOrderedBefore = sort
+    self.orderCriteria = sort
   }
   
   /**
    * Creates a heap from an array. The order of the array does not matter;
    * the elements are inserted into the heap in the order determined by the
-   * sort function.
+   * sort function. For comparable data types, '>' makes a max-heap,
+   * '<' makes a min-heap.
    */
   public init(array: [T], sort: @escaping (T, T) -> Bool) {
-    self.isOrderedBefore = sort
+    self.orderCriteria = sort
     buildHeap(fromArray: array)
   }
   
-  /*
-   // This version has O(n log n) performance.
-   private mutating func buildHeap(array: [T]) {
-   elements.reserveCapacity(array.count)
-   for value in array {
-   insert(value)
-   }
-   }
-   */
-  
   /**
-   * Converts an array to a max-heap or min-heap in a bottom-up manner.
+   * Creates the max-heap or min-heap from an array, in a bottom-up manner.
    * Performance: This runs pretty much in O(n).
    */
-  fileprivate mutating func buildHeap(fromArray array: [T]) {
-    elements = array
-    for i in stride(from: (elements.count/2 - 1), through: 0, by: -1) {
-      shiftDown(i, heapSize: elements.count)
+  private mutating func buildHeap(fromArray array: [T]) {
```

Comment:
This signature could also use some touch up.

I suggest `buildHeap(from array: [T])`. Also thought of changing the name to `configureHeap(from array: [T])`.

The statement "buildHeap" suggests we're creating a new heap, but what is really going on is we're assigning an array to `nodes`, and structuring it as a heap.